examples/step-45/step-45.cc

```diff
@@ -14,417 +14,728 @@
-              << triangulation.n_active_cells ()
-              << std::endl
-              << "Degrees of freedom: " << dof_handler.n_dofs ()
-              << std::endl;
-
-    // Now it is the time for the constraints that come from the periodicity
-    // constraints. We do this in the following, separate function, after
-    // clearing any possible prior content from the constraints object:
-    constraints.clear ();
-    make_periodicity_constraints ();
-
-    // We also incorporate the homogeneous Dirichlet boundary conditions on
-    // the upper and lower parts of the boundary (i.e. the ones with boundary
-    // indicator 1) and close the ConstraintMatrix object:
-    VectorTools::interpolate_boundary_values (dof_handler, 1,
-                                              ZeroFunction<2> (),
-                                              constraints);
+
+    std::vector<unsigned int> block_component (dim+1,0);
+    block_component[dim] = 1;
+    DoFRenumbering::component_wise (dof_handler, block_component);
+
+    std::vector<types::global_dof_index> dofs_per_block (2);
+    DoFTools::count_dofs_per_block (dof_handler, dofs_per_block, block_component);
+    const unsigned int n_u = dofs_per_block[0],
+                       n_p = dofs_per_block[1];
+
+    {
+      owned_partitioning.clear();
+      IndexSet locally_owned_dofs = dof_handler.locally_owned_dofs();
+      owned_partitioning.push_back(locally_owned_dofs.get_view(0, n_u));
+      owned_partitioning.push_back(locally_owned_dofs.get_view(n_u, n_u+n_p));
+
+      relevant_partitioning.clear();
+      IndexSet locally_relevant_dofs;
+      DoFTools::extract_locally_relevant_dofs (dof_handler, locally_relevant_dofs);
+      relevant_partitioning.push_back(locally_relevant_dofs.get_view(0, n_u));
+      relevant_partitioning.push_back(locally_relevant_dofs.get_view(n_u, n_u+n_p));
+
+      constraints.clear ();
+      constraints.reinit(locally_relevant_dofs);
+
+      FEValuesExtractors::Vector velocities(0);
+      FEValuesExtractors::Scalar pressure(dim);
+
+      DoFTools::make_hanging_node_constraints (dof_handler,
+                                               constraints);
+      VectorTools::interpolate_boundary_values (dof_handler,
+                                                0,
+                                                BoundaryValues<dim>(),
+                                                constraints,
+                                                fe.component_mask(velocities));
+      VectorTools::interpolate_boundary_values (dof_handler,
+                                                1,
+                                                BoundaryValues<dim>(),
+                                                constraints,
+
+                                                fe.component_mask(velocities));
+// At this point we need to think about how we want to prescribe
+// periodicity. A face $face_2$  on the left boundary should be matched to
```

Comment:
This doesn't print nicely (and didn't in the introduction) since `$face_2` is rendered as separate variables 'f' 'a' 'c' 'e'. I think you probably want something like `$\text{face}_2$` and similarly elsewhere.